File: mat2fasta.py
##################################################################################
import pandas as pd
import Bio.SeqIO
import argparse
import numpy as np

#functiom to read IUPAC matrix and store as data frame############################
def readmat(mat):
	m=pd.read_csv(mat, delimiter=' ')
	return(m)

#function to read file of additional genome- refpos, ref, alt#####################
def readsnps(snps):
	m=pd.read_csv(snps, delimiter=' ')
	m=m.loc[:,['refpos','alt']]
	return(m)

#function to return union of postions from IUPAC matrix and additional genomes####
def unionpos(mat, nucmersnps):
	matpos=mat.ix[:,'POS'].tolist()
	nucmersnpspos=nucmersnps.iloc[:,0].tolist()
	union=matpos+nucmersnpspos
	union=sorted(set(union))
	return(union)

#iterate unionpos over multiple additional genomes (snps only)####################
def iterateunionpos(mat, additional_snps):
	union=[]
	for additional_snp in additional_snps:
		additional_snp=readsnps(additional_snp)
		union=union+unionpos(mat, additional_snp)
	union=sorted(set(union))
	return(union)

#function to generate new IUPAC row for monomorphic snps##########################
def lookupnewrowIUPAC(mat, reference, pos):
	ref=reference.seq[pos]
	alt='NaN'
	newrow=[pos, ref, alt, 'SNP'] + [ref] * (len(mat.columns) - 4)
	newrow=pd.DataFrame(newrow)
	newrow=newrow.transpose()
	return(newrow)

#function to generate new binary row for monomorphic snps#########################
def lookupnewrowBIN(mat, reference, pos):
	ref=reference.seq[pos]
	alt='NaN'
	newrow=[pos, ref, alt, 'SNP'] + [0] * (len(mat.columns) - 4)
	newrow=pd.DataFrame(newrow)
	newrow=newrow.transpose()
	return(newrow)

#function to return IUPAC matrix adding new positions#############################
def unionposmatIUPAC(mat, reference, additionals_snps):
	mat=readmat(mat)
	reference=Bio.SeqIO.read(reference, 'fasta')
	unionp=iterateunionpos(mat, additionals_snps)
	mpos=mat['POS']
	newpos=set(unionp).difference(set(mpos))
	temp=open('temp.txt', 'w')
	mat.to_csv(temp, mode='w', index=False)
	for pos in newpos:
		newrow=lookupnewrowBIN(mat, reference, pos)
		newrow.to_csv(temp, mode='a', header=False, index=False)
	temp.close()
	mat=pd.read_csv('temp.txt', low_memory=False)
	mat=mat.sort_values(by=['POS'])
	return(mat)


#function to return binary matrix adding new positions############################
def unionposmatBIN(mat, reference, additionals_snps):
	mat=readmat(mat)
	reference=Bio.SeqIO.read(reference, 'fasta')
	unionp=iterateunionpos(mat, additionals_snps)
	mpos=mat['POS']
	newpos=set(unionp).difference(set(mpos))
	temp=open('temp.txt', 'w')
	mat.to_csv(temp, mode='w', index=False)
	for pos in newpos:
		newrow=lookupnewrowBIN(mat, reference, pos)
		newrow.to_csv(temp, mode='a', header=False, index=False)
	temp.close()
	mat=pd.read_csv('temp.txt', low_memory=False)
	mat=mat.sort_values(by=['POS'])
	return(mat)




#####BELOW REQUIRES DEBUGGING
#functions to add snps to iupac and add new IUPAC to matrix as new column##########
#def additionalsIUPAC(mat, additional_snps):
#	for additional_snp in additional_snps:
#		name=additional_snp
#		additional_snp=readsnps(additional_snp)
#		mpos=mat['POS']
#		npos=additional_snp.iloc[:,0]
#		newpos=set(mpos).difference(set(npos))
#		mydict={}
#		for i in  range(len(mat['POS'])):
#			mydict[mat.loc[i, 'POS']]= mat.loc[i, 'REF']
#		newdata=[]
#		for pos in newpos:
#			ref=mydict[pos]
#			newrow=[pos,ref]
#			newdata.append(newrow)
#		newdata=pd.DataFrame(newdata)
#		newdata.columns=['refpos','alt']
#		additional_snp=pd.concat([additional_snp, newdata], axis=0)
#		additional_snp=additional_snp.sort_values(by=['refpos'])
#		newmatcol=pd.DataFrame(additional_snp.iloc[:,1])
#		newmatcol.columns=[name]
#		#check mat['POS']==additional_SNPS['refpos']### STILL TO DO###
#		mat=pd.concat([mat, newmatcol], axis=1)
#	return(mat)


#functions to add additional references to BINARY matrix############################
#def additionalsBIN(mat, additional_snps):
#	for additional_snp in additional_snps:
#		name=additional_snp
#		additional_snp=readsnps(additional_snp)
#		mpos=mat['POS']
#		npos=additional_snp.iloc[:,0]
#		newpos=set(mpos).difference(set(npos))
#		mydict={}
#		for i in  range(len(mat['POS'])):
#			mydict[mat.loc[i, 'POS']]= mat.loc[i, 'REF']
#		newcol=[]
#		for pos in additional_snp.iloc[:,0]:
#			newcol.append(1)
#		additional_snp=pd.concat([additional_snp, pd.DataFrame(newcol)], axis=1)
#		additional_snp.columns=['refpos','alt', 'bin' ]
#		newdata=[]
#		for pos in newpos:
#			ref=mydict[pos]
#			newrow=[pos,ref,0]
#			newdata.append(newrow)
#		newdata=pd.DataFrame(newdata)
#		newdata.columns=['refpos','alt', 'bin' ]
#		additional_snp=pd.concat([additional_snp, newdata], axis=0)
#		additional_snp=additional_snp.sort_values(by=['refpos'])
#		newmatcol=pd.DataFrame(additional_snp.iloc[:,2])
#		newmatcol.columns=[name]
#		#check mat['POS']==additional_SNPS['refpos']### STILL TO DO####
#		mat=pd.concat([mat, newmatcol], ignore_index=True, axis=1)
#	return(mat)



#def additionalsIUPAC(mat, additional_snps):
#	for additional_snp in additional_snps:
#		name=additional_snp
#		additional_snp=readsnps(additional_snp)
#		for pos in mat.loc['POS']:
#			newcol.iloc[pos]=additional_snp.iloc[np.where(additional_snp.loc['REFPOS'] == pos, 'ALT')]
#			except
#				np.where(additional_snp.loc['REFPOS'] == pos)= NA:
#					newcol.iloc[pos]=mat.loc[pos, 'REF']

#	return(mat)
#generate multifasta from IUPAC matrix##############################################
#def matfasta(mat, out):
#	outfasta=open(out, 'w')
#	for col in mat.columns()[3:]:
#		a=mat[col]
#		Bio.SeqIO.write(fasta, outfasta, "fasta")
#	outfasta.close()


###################################################################################
###arg input##
#parser=argparse.ArgumentParser(description='')
#parser.add_argument('--out_type',help='output type required; f (fasta), b \
#	(matbin), i (matiupac)')
#parser.add_argument('--mat', help="Variant matrix in IUPAC format (with 4 d\
#	escriptor columns;'POS' 'REF' 'AL' 'TYPE')")
#parser.add_argument('--matbin', help="Binary variant matrix (with 4 descriptor \
#	columns;'POS' 'REF' 'AL' 'TYPE')")
#parser.add_argument('--out_prefix', help='Prefix for output files')
#parser.add_argument('--additional_snps', help='List of SNP only nucmer files of \
#	additional genomes generated using show-snps', nargs='+')
#parser.add_argument('--reference', help='Reference fasta file')
#args=parser.parse_args()

##pipeline execution###############################################################
#if args.out_type=='i':
#	outmatIUPAC=unionposmatIUPAC(args.mat, args.reference, args.additional_snps)
#	outmat=additionalsIUPAC(outmatIUPAC, additional_snps)
#	a.write(outmat)
#	a.close()m


#if args.out_type=='b':
#	outmatBIN=unionposmatBIN(args.matbin, args.reference, args.additional_snps)
#	outmat=additionalsBIN(outmatBIN, additional_snps)
#	b=open(args.out_prefix + 'mat.bin', 'w')
#	b.write(outmat)
#	b.close()


#if args.out_type=='f':
#	outmat=unionposmatIUPAC(args.mat, args.reference, args.additional_snps)
#	a.write(outmat)
#	a.close()
#	matfasta(args.outmat, args.out_prefix + '.fasta')

##################################################################################
##################################################################################

File: mat2fasta.R
#################################################################################################
library(data.table)
library(seqinr)
library(phylotools)

#preprocessing 
getmat2fastainput<-function(nucmersnps, out_prefix)
{
	snps<-read.delim(nucmersnps, header=F)
	snps2<-as.data.frame(cbind(refpos=as.character(snps[,1]), ref=as.character(snps[,2]), alt=as.character(snps[,3])))
	#deal with duplicate positions?
	write.table(snps2, paste(out_prefix, '_refpos_ref_alt.txt', sep=''), quote=F, row.name=F, col.name=T)
}

#functiom to read IUPAC matrix and store as data frame###########################################
readMat<-function(mat)
{
	mat<-as.data.frame(fread(mat, header=T))
	return(mat)
}

#function to read file of additional genome- refpos, ref, alt####################################
readsnps<-function(snps)
{
	m<-read.table(snps, header=T)
	m<-cbind(refpos=m$refpos,alt=as.character(m$alt))
	return(m)
}
#function to return union of postions from IUPAC matrix and additional genomes###################
unionpos<-function(mat, nucmersnps)
{
	matpos<-mat$POS
	nucmersnpspos=nucmersnps[,1]
	union=sort(unique(c(matpos,nucmersnpspos)))
	return(union)
}

#iterate unionpos over multiple additional genomes (snps only)###################################
iterateunionpos<-function(mat, additional_snps)
{
	union=c()
	for(additional_snp in 1:length(additional_snps))
	{
		additional_snp<-readsnps(additional_snps[additional_snp])
		union<-c(union,unionpos(mat, additional_snp))
	}
	union<-sort(unique(union))
	return(union)
}

#function to generate new IUPAC row for monomorphic snps#########################################
lookupnewrowIUPAC<-function(mat, reference, pos)
{
	ref<-reference[pos]
	alt<-'NA'
	newrow<-as.vector(unlist(c(pos, ref, alt, 'SNP', rep(ref, (length(colnames(mat)) - 4)))))
	return(newrow)
}

#function to generate new binary row for monomorphic snps#########################################
lookupnewrowBIN<-function(mat, reference, pos)
{
	ref<-reference[pos]
	alt<-'NA'
	newrow<-as.vector(unlist(c(pos, ref, alt, 'SNP', rep(0, (length(colnames(mat)) - 4)))))
	return(newrow)
}

#add refernce isolate to IUPAC matrix using REF column
addReferenceIUPAC<-function(mat, reference)
{
	out<-paste(mat, reference, sep='.')
	#mat<-readMat(mat)
	name<-reference
	mat<-cbind(mat, name=mat$REF)
	colnames(mat)[ncol(mat)]<-name
	return(mat)
	#write.table(mat, out, col.name=TRUE, row.name=FALSE, quote=FALSE)

}

#add refernce isolate to bin matrix using REF column
addReferenceBIN<-function(mat, reference)
{
	out<-paste(mat, reference, sep='.')
	#mat<-readMat(mat)
	name<-reference
	mat<-cbind(mat, name=rep(0, nrow(mat)))
	colnames(mat)[ncol(mat)]<-name
	return(mat)
	#write.table(mat, out, col.name=TRUE, row.name=FALSE, quote=FALSE)
}

#merge matrices with same columns but differing positions together
mergeMats<-function(mat1, mat2, out)
{
	m1<-readMat(mat1)
	m2<-readMat(mat2)
	m<-rbind(m1, m2)
	new<-m[order(m$POS),]
	dim(new)
	new[1:20,1:20]
	write.table(new, out, quote=F, row.name=F, col.name=T)
}

##########################################################################################################
#function to return IUPAC matrix adding new positions#####################################################
unionposmatIUPAC<-function(matfile, reference, additionals_snps)
{
	mat<-readMat(matfile)
	reference<-read.fasta(reference, seqonly=TRUE)
	reference<-strsplit(unlist(reference), '')[[1]]
	unionp<-iterateunionpos(mat, additionals_snps)
	mpos<-mat$POS
	newpos<-unionp[is.na(match(unionp, mpos))]
	newpos<-unique(newpos)
	matrix<-matrix(, nrow=length(newpos), ncol= ncol(mat))
	df<-data.frame(matrix)
	colnames(df)<-colnames(mat)
	for(i in 1:length(newpos))
	{
		newrow<-lookupnewrowIUPAC(mat, reference, as.numeric(newpos[i]))
		df[i,]<-newrow
	}
	newmat<-rbind(mat, df)
	newmat<-newmat[order(as.numeric(newmat$POS)),]
	write.table(newmat, paste(matfile, 'extended.temp', sep='_'), col.name=TRUE, row.name=FALSE, quote=FALSE)
	return(newmat)
}

#function to return binary matrix adding new positions#####################################################
unionposmatBIN<-function(matfile, reference, additionals_snps)
{
	mat<-readMat(matfile)
	reference<-read.fasta(reference, seqonly=TRUE)
	reference<-strsplit(unlist(reference), '')[[1]]
	unionp<-iterateunionpos(mat, additionals_snps)
	mpos<-mat$POS
	newpos<-unionp[is.na(match(unionp, mpos))]
	newpos<-unique(newpos)
	headers<-colnames(mat)
	matrix<-matrix(, nrow=length(newpos), ncol= ncol(mat))
	df<-data.frame(matrix)
	colnames(df)<-colnames(mat)
	for(i in 1:length(newpos))
	{
		newrow<-lookupnewrowBIN(mat, reference, as.numeric(newpos[i]))
		df[i,]<-newrow
	}
	mat<-rbind(mat, df)
	mat<-mat[order(df$POS),]
	colnames(mat)<-headers
	write.table(mat, paste(matfile, 'extended.temp', sep='_'), col.name=TRUE, row.name=FALSE, quote=FALSE)
	return(mat)
}

#add isolates to IUPAC matrix containing all reference positions plus more
addIsolatesIUPAC<-function(extended_mat, reference_file, additionals_snps, add_reference=TRUE)
{
	out<-paste(extended_mat, '.additionals', sep='')
	mat<-readMat(extended_mat)
	reference<-read.fasta(reference_file, seqonly=TRUE)
	reference<-strsplit(unlist(reference), '')[[1]]
	for(i in 1:length(additionals_snps))
	{
		name<-additionals_snps[i]
		newisolate<-as.data.frame(readsnps(additionals_snps[i]))
		#next step removes duplicated positions (presumably insertion positions)
		newisolate<-newisolate[-which(duplicated(newisolate$refpos)==TRUE),]
		positions<-mat$POS
		missing_positions<-positions[(is.na(match(positions, newisolate$refpos)))]
		newpositions<-c()
		IUPACcol<-c()
		for(j in 1:length(missing_positions))
		{
			newpositions[j]<-missing_positions[j]
			IUPACcol[j]<-reference[missing_positions[j]]
		}
		newposdf<-as.data.frame(cbind(refpos=as.vector(newpositions), alt=as.character(IUPACcol)))
		origposdf<-as.data.frame(cbind(refpos=as.vector(newisolate$refpos), alt=as.character(newisolate$alt)))
		newdf<-rbind(newposdf, origposdf)
		newdf<-newdf[order(as.numeric(as.character(newdf$refpos))),]
		mat<-cbind(mat, name=newdf$alt)
		colnames(mat)[ncol(mat)]<-name
	}
	if(add_reference==TRUE)
	{
		mat<-addReferenceIUPAC(mat, reference_file)
	}
	write.table(mat, out, col.name=TRUE, row.name=FALSE, quote=FALSE)
}

#add isolates to bin matrix containing all reference positions plus more
addIsolatesBIN<-function(extended_mat,reference_file, additionals_snps, add_reference=TRUE)
{
	out<-paste(extended_mat, '.additionals', sep='')
	mat<-readMat(extended_mat)
	reference<-read.fasta(reference_file, seqonly=TRUE)
	reference<-strsplit(unlist(reference), '')[[1]]
	for(i in 1:length(additionals_snps))
	{
		name<-additionals_snps[i]
		newisolate<-as.data.frame(readsnps(additionals_snps[i]))
		positions<-mat$POS
		missing_positions<-positions[is.na(match(positions, newisolate$refpos))]
		newpositions<-c()
		BINcol<-c()
		for(i in 1:length(missing_positions))
		{
			newpositions[i]<-missing_positions[i]
			BINcol[i]<-0
		}
		newposdf<-as.data.frame(cbind(refpos=as.vector(newpositions), alt=as.character(BINcol)))
		origposdf<-as.data.frame(cbind(refpos=as.vector(newisolate$refpos), alt=as.character(newisolate$alt)))
		newdf<-rbind(newposdf, origposdf)
		newdf<-newdf[order(as.numeric(as.character(newdf$refpos))),]
		mat<-cbind(mat, name=newdf$bin)
		colnames(mat)[ncol(mat)]<-name
	}
	if(add_reference==TRUE)
	{
		mat<-addReferenceBIN(mat, reference_file)
	}
	write.table(mat, out, col.name=TRUE, row.name=FALSE, quote=FALSE)
}


IUPAC2fasta<-function(additionals_mat)
{
	mat<-readMat(additionals_mat)
	mat_snp<-mat[which(mat$TYPE=='SNP'),]
	out<-paste(additionals_mat, '.fasta', sep='')
	for(i in 5:ncol(mat_snp))
	{
		seq<-mat_snp[,i]
		name<-colnames(mat_snp)[i]
		write.fasta(seq,name, out, open='a')
	}
}

FASTA2phylip<-function(fasta_file)
{
	dat<-read.fasta(fasta_file)
	dat2phylip(dat, paste(fasta_file, '.phy',sep=''))
}
	


########################################################################################################################
#### DECIDE AT WHICH POINT TO FILTER OUT INDELS FROM MAT- MAYBE NOT JUST AT LST FASTA GENERATION STEP?
#### FIGURE OUT WHY THERE IS AN X IN THE FASTA FILES
# write shell script to run nucmer to include in repository
